proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,733 @@
+bindings), hence the need for the recursive lookup.
+
+Imports
+^^^^^^^
+
+Normal imports can only appear at the top level of a file. The target of a 
+normal import is a *dotted* name, which is resolved using the
+current rules for locating external Haskell modules.
+
+Importing a module brings all the exported bindings of the target module
+into scope, as described in [hsmods]_. It also brings
+the target module into scope bound to its declared simple name (which may be
+a dotted name if it is a file module).
+
+If a module is imported ``qualified`` then the exported bindings of the target
+module are not brought into scope.
+
+If a module is imported ``as`` a simple name, then the target module is brought
+into scope bound to that simple name instead.
+
+Module export specifiers
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The behaviour of the ``module`` export specifiers are changed as follows.
+
+- ``module M`` *changes* to exports the name ``M`` as a module binding.
+- ``module M(a, b, .. c)`` exports ``M`` and the listed in-scope bindings.
+- ``module M(..)`` exports ``M`` and all of its exported bindings. [#]_
+
+.. [#] See `Do we need additional syntax for re-exporting all bindings from a module?`_ for more discussion.
+
+Backwards compatibility extensions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Flat module exports
+^^^^^^^^^^^^^^^^^^^
+
+The extension ``FlatModuleExports`` is added and enabled by default. It has the effect
+that the ``module M`` export specifier has the behaviour given in [hs2010]_. [#]_
+
+.. [#] See `Flat module exports - reasoning`_ for reasoning.
+
+Import module merging
+^^^^^^^^^^^^^^^^^^^^^
+
+The extension ``ImportModuleMerging`` is added and enabled by default. It has the effect that
+when multiple modules are imported ``as`` the same name, we merge the modules. [#]_
+
+.. [#] See `Import module merging - reasoning`_ for reasoning. 
+   multiple modules are defined with the same name?`_ for discussion
+   about what the primary behaviour should be.
+
+Privileged local module access
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The extension ``PrivilegedLocalModuleAccess`` is added and enabled by default. It has the effect
+that for every file module ``M``, we add a new module declaration with the same name inside ``M``,
+which exports all and only the local declarations of ``M``. [#]_
+
+.. [#] See `Privileged local module access - reasoning`_ for reasoning.
+
+Additional features
+~~~~~~~~~~~~~~~~~~~
+
+The following extensions add features beyond existing Haskell, building upon the previous changes.
+
+Module imports
+^^^^^^^^^^^^^^
+
+We add the language extension ``ImportModule``, which introduces a a new
+syntax ``import module`` [#]_ for importing modules in other contexts.
+The target of an ``import module`` statement is a *qualified* name, resolved
+as normal for qualified names, and will *not* resolve to an external module as
+is possible for normal ``import``s. [#]_ . ``import module``
+otherwise behaves as ``import``.
```

Comment:
I don't think this is exactly what you mean. Here is an example:

```hs
module Top where

import qualified A

import module A.Local (..)
```

The key question: is `Local` in scope here? A normal `import` brings the named module into scope, but I don't think you want `import module` to do that. (That is, I *don't* think `Local` should come into scope.)

Reply:
Why not? I think it should behave like `import` as much as possible. I'd rather take a slightly odd piece of behaviour than the inconsistency.

Reply:
Consider

```hs
module Top where

import Local     -- some top-level module perversely named "Local"
import A

import module A.Local (..)
```

Now, any `Local.` prefix is ambiguous. I could use `as` to rename the `Local` top-level module, but I don't want to, as I really don't need `A.Local` in scope at all.

More generally, I wish Haskell had better ways of bringing components of structures into scope without bringing the structures themselves.

Maybe `import module A.Local (..) hiding (Local)` would do it.

Reply:
`import module A.Local as SomethingElse`?

I don't disagree that this is annoying, but I think this is an annoyance about the way `import` works, which I think is orthogonal to this proposal.